### ISSUE TITLE
Feature request: add NEAT as an optional Trainer optimizer integration

### ISSUE THREAD

**Issue by ItCodinTime** [huggingface/transformers#48541]
### Feature request

I would like to propose optional Hugging Face Trainer integration for NEAT, Nash-Equilibrium Adaptive Training.

NEAT is a conflict-aware optimizer that detects directional disagreement between the current gradient and a stateful optimization signal and selectively reduces the conflicting projection.

Reference implementation:

https://github.com/ItCodinTime/NEAT

NEAT is currently listed in the official Keras ecosystem:

https://keras.io/getting_started/ecosystem/

The proposed user experience would be similar to other optional third-party optimizer integrations:

```python
training_args = TrainingArguments(
    ...,
    optim="neat",
    learning_rate=1e-4,
    optim_args="alpha=0.25,conflict_threshold=0.0",
)

### Motivation

NEAT, Nash Equilibrium Adaptive Training, is a conflict aware optimizer designed for training regimes where meaningful directional disagreement appears between the current gradient and a stateful optimization signal.






Most commonly used optimizers such as AdamW and SGD adapt learning rates, momentum, or gradient scale, but they do not explicitly measure whether the current gradient is directionally opposed to the optimization history. NEAT detects this conflict and selectively reduces the conflicting component while leaving aligned updates unchanged.

The motivation for adding NEAT as an optional Trainer integration is to make conflict aware optimization easier to study in Transformer workloads, particularly in settings such as multi task learning, multi domain fine tuning, competing objectives, noisy optimization, and other regimes where gradient disagreement may be structurally meaningful.

NEAT is not intended as a universal replacement for AdamW or SGD. Its purpose is to provide an explicit mechanism for studying and correcting directional conflict when that conflict is actually present.

The reference implementation is available here:

https://github.com/ItCodinTime/NEAT

NEAT is also listed in the official Keras ecosystem as a Keras first optimizer library for conflict aware neural network training:

https://keras.io/getting_started/ecosystem/

Transformers already provides a convenient interface for specialized optimizer integrations through `TrainingArguments.optim`. A NEAT integration could provide the same user experience while keeping the optimizer implementation in the external NEAT package.

For example:

```python
training_args = TrainingArguments(
    ...,
    optim="neat",
    learning_rate=1e-4,
    optim_args="alpha=0.25,conflict_threshold=0.0",
)

### Your contribution


### Your contribution

```markdown
Yes. I would be happy to implement the integration and submit the PR.

I can contribute the complete Transformers side of the integration, following the existing patterns used for optional third party optimizers.

Specifically, I can add:

- a `NEAT` entry to the supported optimizer names
- lazy optional import of the external NEAT package
- optimizer construction through `Trainer`
- support for passing NEAT specific parameters through `optim_args`
- a clear dependency error with installation instructions when NEAT is not installed
- documentation in the Transformers optimizer guide
- example `TrainingArguments` usage
- integration and regression tests
- any formatting, typing, or documentation updates required by the Transformers contribution guidelines

I maintain the NEAT reference implementation and can also make any package side API changes needed to keep the Transformers integration small and consistent with existing optimizer integrations.

I am happy to adjust the proposed API based on maintainer feedback before submitting the implementation PR.
